brave/src/main/java/brave/propagation/ExtraFieldPropagation.java

```diff
@@ -277,11 +409,28 @@ public static void set(TraceContext context, String name, String value) {
     final String[] fieldNames;
     final String[] keyNames;
     final int[] keyToField;
-    final BitSet redacted;
     final ExtraFactory extraFactory;
+    final Plugin[] plugins;
 
-    Factory(Propagation.Factory delegate, String[] fieldNames, String[] keyNames, BitSet redacted) {
-      this(delegate, fieldNames, keyNames, keyToField(keyNames), redacted);
+    @SuppressWarnings("ThreadLocalUsage") // intentional: instances may have different plugin counts
+    final ThreadLocal<Object[]> FIELD_UPDATERS = new ThreadLocal<>();
```

Comment:
Hmm, can you explain this further (even in just a reply here). I don't follow why a thread local is necessary here, I must be missing some context

Reply:
otherwise we allocate one throw-away array for each incoming request and each outgoing one, on top of the other overhead added by this feature. This is an attempt to reduce the garbage impact.

Reply:
Anyway good question. basically I used to only put in features with one thing like one `Reporter<Span>`, which you wouldn't need to loop over. However in plugins, this was designed for multiple and we can be a bit more efficient than end users are likely to be when trying to compose them.

I will cleanup this anyway, for example only using it when there are in fact multiple plugins present. that will help isolate the benchmarking concern, too, and protect the unsuspecting reviewer from seeing too much in one file, too!